File: objectivec/day11_part1_2015.m
#import <Foundation/Foundation.h>

// Function to increment a character
char incrementChar(char c) {
    if (c == 'z') return 'a';
    return c + 1;
}

// Function to check if a string contains at least one increasing straight of at least three letters
BOOL hasIncreasingStraight(NSString *str) {
    for (NSInteger i = 0; i < str.length - 2; i++) {
        if ([str characterAtIndex:i] == [str characterAtIndex:i + 1] - 1 &&
            [str characterAtIndex:i + 1] == [str characterAtIndex:i + 2] - 1) {
            return YES;
        }
    }
    return NO;
}

// Function to check if a string contains at least two different, non-overlapping pairs of letters
BOOL hasTwoPairs(NSString *str) {
    NSInteger pairs = 0;
    for (NSInteger i = 0; i < str.length - 1; i++) {
        if ([str characterAtIndex:i] == [str characterAtIndex:i + 1]) {
            pairs++;
            i++; // Skip the next character
        }
    }
    return pairs >= 2;
}

// Function to check if a string contains any of the forbidden letters
BOOL hasForbiddenLetters(NSString *str) {
    for (NSInteger i = 0; i < str.length; i++) {
        char c = [str characterAtIndex:i];
        if (c == 'i' || c == 'l' || c == 'o') {
            return YES;
        }
    }
    return NO;
}

// Function to increment a string
NSString *incrementString(NSString *str) {
    NSMutableString *result = [NSMutableString stringWithString:str];
    for (NSInteger i = result.length - 1; i >= 0; i--) {
        char c = [result characterAtIndex:i];
        if (c == 'z') {
            [result replaceCharactersInRange:NSMakeRange(i, 1) withString:@"a"];
        } else {
            [result replaceCharactersInRange:NSMakeRange(i, 1) withString:[NSString stringWithFormat:@"%c", incrementChar(c)]];
            return result;
        }
    }
    return result;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        // Read input from file
        NSString *filePath = @"input.txt";
        NSString *input = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        
        // Trim whitespace and convert to lowercase
        input = [input stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        input = [input lowercaseString];
        
        // Find the next valid password
        while (YES) {
            input = incrementString(input);
            if (!hasForbiddenLetters(input) && hasIncreasingStraight(input) && hasTwoPairs(input)) {
                break;
            }
        }
        
        // Print the result
        NSLog(@"%@", input);
    }
    return 0;
}

File: training_data/reflection-tuning/day11_part1_2015.go
package main

import (
	"fmt"
	"strings"
)

func incrementPassword(password string) string {
	pw := []byte(password)
	for i := len(pw) - 1; i >= 0; i-- {
		if pw[i] == 'z' {
			pw[i] = 'a'
		} else {
			pw[i]++
			break
		}
	}
	return string(pw)
}

func hasIncreasingStraight(password string) bool {
	for i := 0; i < len(password)-2; i++ {
		if password[i+1] == password[i]+1 && password[i+2] == password[i]+2 {
			return true
		}
	}
	return false
}

func hasNoForbiddenLetters(password string) bool {
	return !strings.ContainsAny(password, "iol")
}

func hasTwoPairs(password string) bool {
	pairs := 0
	for i := 0; i < len(password)-1; i++ {
		if password[i] == password[i+1] {
			pairs++
			i++ // Skip next character to avoid overlapping
		}
	}
	return pairs >= 2
}

func isValidPassword(password string) bool {
	return hasIncreasingStraight(password) &&
		hasNoForbiddenLetters(password) &&
		hasTwoPairs(password)
}

func nextPassword(current string) string {
	next := incrementPassword(current)
	for !isValidPassword(next) {
		next = incrementPassword(next)
	}
	return next
}

func main() {
	input := "hxbxwxba" // Example input
	fmt.Println(nextPassword(input))
}
